zebra-state/src/service/check/tests/utxo.rs

```diff
@@ -0,0 +1,764 @@
+            commit_result,
+            Err(EarlyTransparentSpend {
+                outpoint: expected_outpoint,
+            }.into())
+        );
+        prop_assert_eq!(Some((Height(0), genesis.hash)), state.best_tip());
+
+        // the non-finalized state did not change
+        prop_assert!(state.mem.eq_internal_state(&previous_mem));
+
+        // the finalized state does not have the UTXO
+        prop_assert!(state.disk.utxo(&expected_outpoint).is_none());
+    }
+}
+
+/// State associated with transparent UTXO tests.
+struct TestState {
+    /// The pre-populated state service.
+    state: StateService,
+
+    /// The genesis block that has already been committed to the `state` service's
+    /// finalized state.
+    #[allow(dead_code)]
+    genesis: FinalizedBlock,
+
+    /// A block at height 1, that has already been committed to the `state` service.
+    block1: Arc<Block>,
+}
+
+/// Return a new `StateService` containing the mainnet genesis block.
+/// Also returns the finalized genesis block itself.
+fn new_state_with_mainnet_transparent_data(
+    inputs: impl IntoIterator<Item = transparent::Input>,
+    outputs: impl IntoIterator<Item = transparent::Output>,
+    use_finalized_state: bool,
+) -> TestState {
+    let (mut state, genesis) = new_state_with_mainnet_genesis();
+    let previous_mem = state.mem.clone();
+
+    let mut block1 = zebra_test::vectors::BLOCK_MAINNET_1_BYTES
+        .zcash_deserialize_into::<Block>()
+        .expect("block should deserialize");
+
+    let outputs: Vec<_> = outputs.into_iter().collect();
+    let outputs_len: u32 = outputs
+        .len()
+        .try_into()
+        .expect("unexpectedly large output iterator");
+
+    let transaction = transaction_v4_with_transparent_data(inputs, outputs);
+    let transaction_hash = transaction.hash();
+
+    let expected_outpoints = (0..outputs_len).map(|index| transparent::OutPoint {
+        hash: transaction_hash,
+        index,
+    });
+
+    block1.transactions[0] = transaction_v4_from_coinbase(&block1.transactions[0]).into();
+    block1.transactions.push(transaction.into());
+
+    let block1 = Arc::new(block1);
+
+    if use_finalized_state {
+        let block1 = FinalizedBlock::from(block1.clone());
+        let commit_result = state.disk.commit_finalized_direct(block1.clone(), "test");
+
+        // the block was committed
+        assert_eq!(Some((Height(1), block1.hash)), state.best_tip());
+        assert!(commit_result.is_ok());
+
+        // the non-finalized state didn't change
+        assert!(state.mem.eq_internal_state(&previous_mem));
+
+        for expected_outpoint in expected_outpoints {
+            // the finalized state has the UTXOs
+            assert!(state.disk.utxo(&expected_outpoint).is_some());
+            // the non-finalized state does not have the UTXOs
+            assert!(state.mem.any_utxo(&expected_outpoint).is_none());
+        }
+    } else {
+        let block1 = block1.clone().prepare();
```

Comment:
Minor: Is the `clone()` redundant?

Reply:
Nope, if it was, clippy would warn and the job would fail.

(It's required because `prepare` takes ownership of the `Arc<Block>`.)